### PR TITLE
Move Map and Plugin object to a tree structure

### PR DESCRIPTION
**Pull request by BibMartin**
3 simple commits, and an important one (https://github.com/python-visualization/folium/commit/561f060f3b17c3039ef708f100be8596bf7ae44f).

The goal of the latter is to move progressively the Map and Plugin objects into a tree structure.

What I have in mind is a tree like :

    +- figure
        +- header
        +- css
        +- body
        +- script
        +- maps
            +- Markers
            +- Layers
            +- Plugins
            ...
        +- Others

Each *node* of the tree would have a `render` method that can have an effect on his parents.

This would have several positive effects:
* One can embed several maps into a single page, and eventually add other html stuff, as asked in #138.
* Each Leaflet object would have a python *mirror* so that one can compute interactions between objects. For example, `marker`, `popup`, `divIcon` and `awesomeMarker` need that kind of interactions (see #169). In my mind, this corresponds to the idea of a `FoliumDataSource`, as @wrobstory mentioned in #34.
* Some plugins will be easier to implement : provided that a Jinja temple has a `render` method, it can be a leaf in that tree. This may help today's plugins and @themiurgo's #147 to converge.
* Each node implements how it's children are rendered (and in which order), which provides more flexibility.
* Issue #155 would find an elegant solution : an external link would be an object that you can render several ways (http, https, embeded). (see https://github.com/python-visualization/folium/commit/5e90c1bf3f2e5e51c062c041393b95c3386710ba)

Well, discussion is open... ;-)

### CODE REVIEW

**Comment by BibMartin** [comment]
I forgot ; for the moment, the tree structure is:

    +- Map
        +- figure
            +- header
            +- css
            +- body
            +- script
        +- template_vars
        +- plugins
        ...

The idea is to move progressively `template_vars` into a structured tree (without changing the API), then to have `figure` embed several maps.

**Comment by themiurgo** [comment]
I feel like this PR is actually 2PRs, one part is the plugin render idea, the other is the tree structure.

Some questions:
- Why is the root name called figure? Shouldn't it be something like "page" or "root"?
- Why is plugin a child of maps? Surely each plugin affects the whole page and needs to inject their own header. If I embed two maps with the same sets of headers it will be difficult to track whether the js/css has been already injected beforehand.

**Comment by BibMartin** [comment]
> Why is the root name called figure? Shouldn't it be something like "page" or "root"?

I chose `figure` because I have something like matplotlib in mind : a figure in which you can put several maps/axes. I thought we could get closer to their function names. I would be okay for "page" or "root".

> Why is plugin a child of maps?

A plugin generally adds something on the map. But you're true, it can require to add somthing to the overall figure. This is why I have designed the thing so that each node knows who is his parent, and can have an effect on it. For example, if you have something like 

    Figure
        map1
            boatmarker1
            boatmarker2
        map2
            boatmarker3
            boatmarker4

Then each boatmarker will do something like

    boatmarker.map.figure.header['boatmarker_js'] = Link('http://...')

But as it is the same Figure object, they will simply overwrite the former header.
BTW, this is why we need to use `OrderedDict` instead of lists : order is important, but we also need to get objects by their key/name.

**Comment by BibMartin** [comment]
Let's say that `VegaPopup` is working (though it's pure prototype). @wrobstory I would be happy if you have time to wartch https://github.com/python-visualization/folium/commit/8ce53ea009654e2f67e1220436b81650d327791c as it is going far from `folium.Map._popup_render` and I'm not really sure of the javascript generated.

**Comment by BibMartin** [comment]
:green_heart: At last !

File: folium/plugins/marker.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+    def render(self, **kwargs):
+        """TODO : docstring here"""
+        assert self.map is not None, "Cannot render a Popup that have no 'map' attribute."
+        self.map.figure.header['d3'    ] = JavascriptLink("https://cdnjs.cloudflare.com/ajax/libs/d3/3.5.5/d3.min.js")
+        self.map.figure.header['vega'  ] = JavascriptLink("https://cdnjs.cloudflare.com/ajax/libs/vega/1.4.3/vega.min.js")
+        self.map.figure.header['jquery'] = JavascriptLink("https://code.jquery.com/jquery-2.1.0.min.js")
+        self.map.figure.script['vega_parse'] = Template("""function vega_parse(spec, div) {
+            vg.parse.spec(spec, function(chart) { chart({el:div}).update(); });}""")
+
```
Comment (ocefpaf) [inline folium/plugins/marker.py]: With this approach it should be relatively easy to create mpld3 (pr even bokeh) popups.

**Comment by BibMartin** [comment]
This is outdated. Replaced by #203

**Pull request closed without merging**